confdgnmi/README.md

```diff
@@ -0,0 +1,283 @@
+`./src/confd_gnmi_client.py -o subscribe --prefix /interfaces --path interface[name=if_82]/name --path interface[name=if_82]/type`
+
+NOTE: Other parameters (username, password, subscription types are currently hardcoded, TODO)
+    
+## Limitations and TODOs of current implementation
+
+* `encoding` - used only `BYTES`
+* `Get`, `Set` and `Subscribe` works only on `leaf` elements
+* `Subscribe` - only `ONCE` and `POLL` is implemented so far
+    * `updates_only` not supported
+    * `heartbeat_interval` not supported
+    * `sync_response` not generated
+* all values `TypedValues` are used as strings (`string_val`)
+* gNMI Path is converted to XPath or formatted path with simple string operations 
+  no datamodel knowledge used) 
+  
+## Investigations/TODOs
+
+TODO converting XPath to ConfD formatted path/keypath and back.
+(Work with `cs_node` for API adapter, for Netconf adapter ??)
+
+## Analysis
+
+We use (part) of `ietf-interfaces.yang` as sample datamodel.
+
+### gNMI interface 
+
+#### Capabilities
+
+`rpc Capabilities(CapabilityRequest) returns (CapabilityResponse);`
+
+`Capability` returns list of `supported_models` and list of `encodings`.
+It can also return list of gNMI extensions.  
+
+Each model has following attributes: 
+ 
+`name` (string), `organization` (string), `version`
+
+##### Implementation 
+
+Capability information can be fetched from `"/ncm:netconf-state/ncm:capabilities/ncm:capability"`  
+datamodel, found in the `ietf-netconf-monitoring.yang`.  
+
+Another option can be to get it with `confd_get_nslist` API call:
+
+Call  `confd_get_nslist`, array of following is returned:
+```
+struct confd_nsinfo {
+    const char *uri;
+    const char *prefix;
+    u_int32_t hash;
+    const char *revision;
+    const char *module;
+}; 
+```
+
+`name` -> `module`
+`organization` -> TODO (cannot get with ConfD API, return empty string)
+`version` -> `revision` (return revision string -date ?)
+
+TODO: Pass yang file content as extension? HOW to get yang files from ConfD?
+
+#### Get
+  
+`rpc Get(GetRequest) returns (GetResponse);`
+
+Pass in request:
+list of `path`, `type` (CONFIG, OPERATIONAL, STATE), `encoding`, 
+list of models to be used (`use_models`)
+
+Get in response list of `notifications`. 
+
+TODOs:
+
+* notification contains list of updated paths with values and deleted paths.
+What should be returned? We return only `updated` paths.
+
+* The notification should contain whole subtree for each request path.
+This means, for each request path there will be list of paths in 
+`update`, each with value.  
+
+* Since there is a path associated with each value, for requests on nested lists
+we have to find out, which elements are keys and add them to the Path(s).
+
+* What value type should be for empty or presence container ??? (bool ???)
+
+##### Implementation 
+
+`maapi_save_config`  - can be used even for operational data (` MAAPI_CONFIG_WITH_OPER`).
```

Comment:
RESTCONF is another option where you can choose XML or JSON

Reply:
RESTCONF can be added in next milestones